### PR TITLE
INC-805 - Get behaviour counts based on last review date

### PR DESCRIPTION
**Pull request by Mjwillis**
## What does this do?

- Calls a new endpoint in prison api which returns a list of case note counts group by type/subtype for a list of prisoners but also filters by a `fromDate` which is the last review date
- This is used to then calculate the number of positive (POS) and negative (NEG) case note behaviours raised since the last review.

### Last Review Date Calculation
The last review date is as follows:
- find the latest `Real Review` (REVIEW/MIGRATED) incentive review.  If found - this is the review date
- if no `Real Review` found then use the review date 3 month before today
- if this review date is over 3 month old then make the review date 3 month before today

### CODE REVIEW

File: src/main/kotlin/uk/gov/justice/digital/hmpps/incentivesapi/service/IncentiveReviewsService.kt
Comment (Mjwillis) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/incentivesapi/service/IncentiveReviewsService.kt:107]: The levelCode was shadowing the review level code

File: src/main/kotlin/uk/gov/justice/digital/hmpps/incentivesapi/service/IncentiveReviewsService.kt
Comment (Mjwillis) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/incentivesapi/service/IncentiveReviewsService.kt:150]: Will also add in the date check if older than 3 month and default to that

File: src/main/kotlin/uk/gov/justice/digital/hmpps/incentivesapi/service/IncentiveReviewsService.kt
Comment (Mjwillis) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/incentivesapi/service/IncentiveReviewsService.kt:155]: Aren't interested in the sub type count anymore

File: src/main/kotlin/uk/gov/justice/digital/hmpps/incentivesapi/service/BehaviourService.kt
Comment (xoen) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/incentivesapi/service/BehaviourService.kt:45]: Are we sure this `?: review.value.first()` is the correct thing to do?

Someone may have been in prison for months/years without a genuine Incentive review.
They may have a number of relevant case notes in the last 3 months.
But if, let's say, they got transferred last week this `review.value.first()` would mean only case notes from the last week would returned, right?

Maybe this should fallback to 3 months ago when no real review is found? Sorry if this is a stupid question, I'm still working my way through this PR.

Comment (Mjwillis) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/incentivesapi/service/BehaviourService.kt:45]: Its a good point. I could change it so that if no real review found then just get three months

File: src/main/kotlin/uk/gov/justice/digital/hmpps/incentivesapi/service/IncentiveReviewsService.kt
Comment (xoen) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/incentivesapi/service/IncentiveReviewsService.kt:67]: I'm not sure how big is the window of opportunity here but potentially someone may have no entries and therefore not be in this list (even tho they may have a some case notes maybe?)

Comment (xoen) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/incentivesapi/service/IncentiveReviewsService.kt:67]: Another thing to bear in mind is that this list could be very very big, especially if the wing has a few prisoners which have been in prison for years and years and have a lot of records.

In theory we're only interested in the last 3 months, but of course adding that to the query means you would remove prisoners without a recent review from the list, and therefore their case notes would be returned.

Comment (Mjwillis) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/incentivesapi/service/IncentiveReviewsService.kt:67]: Every prisoner MUST have an incentive review, in which case they will all be considered -Good point though!

File: src/main/kotlin/uk/gov/justice/digital/hmpps/incentivesapi/service/BehaviourService.kt
Comment (Mjwillis) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/incentivesapi/service/BehaviourService.kt]: @xoen I've updated this to your suggestion

File: src/main/kotlin/uk/gov/justice/digital/hmpps/incentivesapi/service/IncentiveSummaryService.kt
Comment (xoen) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/incentivesapi/service/IncentiveSummaryService.kt:51]: I may miss something but you could reuse the `reviews` value here AFAICT should contain the same list so we can save the trip to the DB.

**Review by xoen — COMMENTED** [review]
LGTM, however this is probably worth changing before merging :)

https://github.com/ministryofjustice/hmpps-incentives-api/pull/360/files/1baec61de94aed920a2dadb3c9bf22ae8e94a298#r1109960720

File: src/main/kotlin/uk/gov/justice/digital/hmpps/incentivesapi/service/IncentiveSummaryService.kt
Comment (Mjwillis) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/incentivesapi/service/IncentiveSummaryService.kt:51]: Duh, I specifically split this out so I could do that, must have been a cut and paste job when refactoring!  Will change

**Comment by Mjwillis** [comment]
@xoen update PR based on your very good suggestions!

**Review by xoen — APPROVED** [review]

**Pull request merged**
